Update typing for pg

aleung opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [ ] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [ ] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

What's changed in this PR:
* Fix definition error in `Client.query` and `Pool.query` when `strictNullChecks` is enable in tsconfig.
* Add lint and fix errors.

sheetalkamat commented on types/pg/index.d.ts:
```diff
@@ -9,9 +9,11 @@ import events = require("events");
+// tslint:disable-next-line unified-signatures
+export function connect(connection: string, callback: (err: Error, client: Client, done: (err?: any) => void) => void): void;
+// tslint:disable-next-line unified-signatures
+export function connect(config: ClientConfig, callback: (err: Error, client: Client, done: (err?: any) => void) => void): void;
```
What is the reason to not write the connect declaration as:
```TypeScript
export function connect(connectionOrConfig: string | ClientConfig, callback: (err: Error, client: Client, done: (err?: any) => void) => void): void;
```

pspeter3 replied:
Yeah, that seems like a better way to express it.

aleung commented:
@sheetalkamat @pspeter3 Please review my update.

pspeter3 commented on types/pg/index.d.ts:
```diff
@@ -9,10 +9,9 @@ import events = require("events");
-// tslint:disable-next-line unified-signatures
-export function connect(config: ClientConfig, callback: (err: Error, client: Client, done: (err?: any) => void) => void): void;
+export function connect(
+    connectionOrConfig: string | ClientConfig,
```
I wish there was a name for this, for example `Configuration`. You could then do the following

```
type Configuration = string | Client config

export function connect(configuration: Configuration
```

You could also use it for the client constructor.

abenhamdine replied:
Pg does not allow anymore to pass directly a connection string to the pool, but allows it as a property of the config object : 
```js
const pool = new Pool({
  connectionString:  'postgresql://dbuser:secretpassword@database.server.com:3211/mydb'
})
```
see https://node-postgres.com/features/connecting#connection-uri

Same thing for client config.
So we should add the config definition : 
```typescript
{
connectionString?: string
}
```

aleung replied:
@abenhamdine Got it. Please review the update.

aleung commented on types/pg/index.d.ts:
```diff
@@ -9,10 +9,9 @@ import events = require("events");
-// tslint:disable-next-line unified-signatures
-export function connect(config: ClientConfig, callback: (err: Error, client: Client, done: (err?: any) => void) => void): void;
+export function connect(
+    connectionOrConfig: string | ClientConfig,
```
When giving a type name `Configuration`, the hint message in IDE will be `function connect(configuration: Configuration)`. It hides the information that `configuration` is either a connection string or a `Config` object. In previous revision I was intended to leave `connect` function separated as overloads to make IDE hint straightforward.

It's just my two cents, from IDE hint perspective. On the other hand, giving a type name follows DRY principle. I can update the code as you advised, please feel free to let me know.

Another thing: I just read pg document on https://node-postgres.com/, and couldn't find any documentation about usage of `connect(connection: string)`. It looks like that pg has deprecated this usage and only leaves `connect(config: ClientConfig)`. I don't know whether the typing should exposing undocumented usage?

aleung commented:
I've made a new commit to update typing to align with pg@7.1 API. Removed deprecated functions. Updated test with examples from pg official document at https://node-postgres.com/

@sheetalkamat @pspeter3 - Could you please take some time to review it?

abenhamdine commented on types/pg/index.d.ts:
```diff
@@ -73,31 +70,33 @@ export declare class Pool extends events.EventEmitter {
-    query(queryText: string, values: any[]): Promise<QueryResult>;
-
+    query(queryConfig: QueryConfig): Promise<QueryResult>;
+    query(queryText: string, values?: any[]): Promise<QueryResult>;
```
It looks like properties totalCount, idleCount, waitingCount are missing, see https://node-postgres.com/api/pool#-code-pool-totalcount-int-code- and

aleung replied:
Added. But there may still be some other missing properties or methods, since the purpose of this PR is to fix definition error rather then make it complete.

aleung commented on types/pg/index.d.ts:
```diff
@@ -40,6 +36,7 @@ export interface PoolConfig extends ClientConfig {
       max?: number;
```
But is it documented? I can't find it from document and example.

abenhamdine replied:
Ah you're right I think I only saw it in the code.
Perhaps we rather avoid to add it, and sticks to max only which is the 'official' property since 7.0

aleung commented on types/pg/index.d.ts:
```diff
@@ -40,6 +36,7 @@ export interface PoolConfig extends ClientConfig {
       refreshIdle?: boolean;
+      connectionTimeoutMillis?: number;
       idleTimeoutMillis?: number;
       reapIntervalMillis?: number;
```
Updated

minestarks commented on types/pg/index.d.ts:
```diff
@@ -1,6 +1,6 @@
-// Type definitions for pg 6.1
+// Type definitions for pg 7.1
```
Could you back up version 6 to another folder as decsribed in the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped#i-want-to-update-a-package-to-a-new-major-version)?

aleung replied:
I have copied the files prior to my change to folder `v6`.
